contracts/staking/stakeManager/StakeManager.sol

```diff
@@ -631,38 +684,288 @@ contract StakeManager is IStakeManager, StakeManagerStorage, Initializable {
+        view
+        returns(UnsignedValidatorsContext memory)
+    {
+        while (context.validatorIndex < totalValidators && context.validators[context.validatorIndex] != signer) {
+            context.unsignedValidators[context.unsignedValidatorIndex] = signerToValidator[context.validators[context.validatorIndex]];
+            context.unsignedValidatorIndex++;
+            context.validatorIndex++;
+        }
+
+        context.validatorIndex++;
+        return context;
+    }
+
+    function _increaseRewardAndAssertConsensus(
+        uint256 blockInterval,
+        address proposer,
+        uint256 signedStakePower,
+        bytes32 stateRoot,
+        uint256[] memory unsignedValidators,
+        uint256 totalUnsignedValidators,
+        uint256[] memory deactivatedValidators,
+        uint256 totalDeactivatedValidators
+    ) private returns (uint256) {
+        uint256 currentTotalStake = validatorState.amount;
+        require(signedStakePower >= currentTotalStake.mul(2).div(3).add(1), "2/3+1 non-majority!");
+
+        // checkpoint rewards are based on BlockInterval multiplied on `CHECKPOINT_REWARD`
+        // for bigger checkpoints reward is capped at `CHECKPOINT_REWARD`
+        // if interval is 50% of checkPointBlockInterval then reward R is half of `CHECKPOINT_REWARD`
+        // and then stakePower is 90% of currentValidatorSetTotalStake then final reward is 90% of R
+        uint256 reward = blockInterval.mul(CHECKPOINT_REWARD).div(checkPointBlockInterval);
+        reward = reward.mul(signedStakePower).div(currentTotalStake);
+        reward = Math.min(CHECKPOINT_REWARD, reward);
+
+        uint256 _proposerBonus = reward.mul(proposerBonus).div(MAX_PROPOSER_BONUS);
+        uint256 proposerId = signerToValidator[proposer];
+
+        Validator storage _proposer = validators[proposerId];
+        uint256 delegatedAmount = _proposer.delegatedAmount;
+        if (delegatedAmount > 0) {
+            _increaseValidatorRewardWithDelegation(proposerId, _proposer.amount, delegatedAmount, _proposerBonus);
+        } else {
+            _proposer.reward = _proposer.reward.add(_proposerBonus);
+        }
+
+        // update stateMerkleTree root for accounts balance on heimdall chain
+        accountStateRoot = stateRoot;
+
+        uint256 newRewardPerStake = rewardPerStake.add(
+            reward.sub(_proposerBonus).mul(REWARD_PRECISION).div(signedStakePower)
+        );
+
+        // evaluate rewards for validator who did't sign and set latest reward per stake to new value to avoid them from getting new rewards.
+        _updateValidatorsRewards(unsignedValidators, totalUnsignedValidators, newRewardPerStake);
+
+        // distribute rewards between signed validators
+        rewardPerStake = newRewardPerStake;
+
+        // evaluate rewards for unstaked validators to avoid getting new rewards until they claim their stake
+        _updateValidatorsRewards(deactivatedValidators, totalDeactivatedValidators, newRewardPerStake);
```

Comment:
Shouldn't `_updateValidatorsRewards` before `rewardsPerStake` or after? 

`_updateValidatorsRewards` for both `unsigned` and `deactivated`.

Reply:
For unstaked - no, they must get the last reward.